File: server/internal/auth/metadata.go
package auth

import "net/http"

func (s *Service) AuthorizationMetadata(w http.ResponseWriter, r *http.Request) {
	issuer := s.IssuerURL()
	metadata := map[string]any{
		"issuer":                                issuer,
		"authorization_endpoint":                issuer + "/authorize",
		"token_endpoint":                        issuer + "/token",
		"response_types_supported":              []string{"code"},
		"grant_types_supported":                 []string{"authorization_code", "refresh_token"},
		"code_challenge_methods_supported":      []string{"S256"},
		"token_endpoint_auth_methods_supported": []string{"none"},
		"scopes_supported":                      AllScopes,
	}
	if s.dcrSettings().Enabled {
		metadata["registration_endpoint"] = issuer + "/register"
	}
	writePublicJSON(w, metadata)
}

func (s *Service) dcrSettings() DCRSettings {
	return s.cfg.DCRSettings()
}

func (s *Service) ProtectedResourceMetadata(w http.ResponseWriter, r *http.Request) {
	writePublicJSON(w, s.resourceMetadata(""))
}

func (s *Service) MCPProtectedResourceMetadata(w http.ResponseWriter, r *http.Request) {
	writePublicJSON(w, s.resourceMetadata("/mcp"))
}

func (s *Service) resourceMetadata(path string) map[string]any {
	issuer := s.IssuerURL()
	return map[string]any{
		"resource":                 issuer + path,
		"authorization_servers":    []string{issuer},
		"scopes_supported":         AllScopes,
		"bearer_methods_supported": []string{"header"},
	}
}

func (s *Service) AuthConfig(w http.ResponseWriter, r *http.Request) {
	cfg := s.config()
	scopes := []string{}
	if cfg.MasterPassword != "" {
		scopes = AllScopes
	}
	writePublicJSON(w, map[string]any{
		"master_password_status": cfg.masterPasswordStatus(),
		"google_auth_enabled":    s.GoogleEnabled(),
		"email_auth_enabled":     s.EmailEnabled(),
		"webauthn_enabled":       s.PassKeyEnabled(),
		"disable_all_auth":       cfg.DisableAllAuth,
		"setup_complete":         s.cfg.SetupComplete(),
		"scopes":                 scopes,
	})
}

File: server-rs/src/auth/oauth/metadata.rs
use std::sync::Arc;

use axum::{extract::State, response::Response};

use super::super::{ALL_SCOPES, Service};
use super::public_json;

pub(super) async fn authorization_metadata(State(service): State<Arc<Service>>) -> Response {
    let issuer = service.issuer_url();
    let mut metadata = serde_json::json!({
        "issuer": issuer,
        "authorization_endpoint": format!("{issuer}/authorize"),
        "token_endpoint": format!("{issuer}/token"),
        "response_types_supported": ["code"],
        "grant_types_supported": ["authorization_code", "refresh_token"],
        "code_challenge_methods_supported": ["S256"],
        "token_endpoint_auth_methods_supported": ["none"],
        "scopes_supported": scope_names(),
    });
    if service.dcr_settings().enabled {
        metadata["registration_endpoint"] = format!("{issuer}/register").into();
    }
    public_json(metadata)
}

pub(super) async fn protected_resource_metadata(State(service): State<Arc<Service>>) -> Response {
    public_json(resource_metadata(&service, ""))
}

pub(super) async fn mcp_protected_resource_metadata(State(service): State<Arc<Service>>) -> Response {
    public_json(resource_metadata(&service, "/mcp"))
}

pub(super) async fn auth_config(State(service): State<Arc<Service>>) -> Response {
    let scopes = service
        .master_password()
        .is_some()
        .then(scope_names)
        .unwrap_or_default();
    public_json(serde_json::json!({
        "master_password_status": service.master_password_status().to_string(),
        "google_auth_enabled": service.google_enabled(),
        "email_auth_enabled": service.email_enabled(),
        "webauthn_enabled": service.passkey_enabled(),
        "disable_all_auth": service.disable_all_auth(),
        "setup_complete": service.setup_complete(),
        "scopes": scopes,
    }))
}

fn resource_metadata(service: &Service, path: &str) -> serde_json::Value {
    let issuer = service.issuer_url();
    serde_json::json!({
        "resource": format!("{issuer}{path}"),
        "authorization_servers": [issuer],
        "scopes_supported": scope_names(),
        "bearer_methods_supported": ["header"],
    })
}

fn scope_names() -> Vec<String> {
    ALL_SCOPES.iter().map(ToString::to_string).collect()
}
